### ISSUE TITLE
[DRAFT] Next phase of neural_network.cpp

### ISSUE THREAD

**Issue by christimahu** [christimahu/blahajpi#13]
#include <vector>
#include <cmath>
#include <random>
#include <iostream>
#include <stdexcept>
#include <numeric>

namespace blahajpi {
namespace models {

class NeuralNetworkClassifier {
public:
    NeuralNetworkClassifier(int inputSize, int hiddenSize, int outputSize, int epochs, double learningRate)
        : inputSize_(inputSize), hiddenSize_(hiddenSize), outputSize_(outputSize),
          epochs_(epochs), learningRate_(learningRate) {

        // Xavier initialization for weights
        std::random_device rd;
        std::mt19937 gen(rd());
        std::normal_distribution<> dist(0, 1.0 / std::sqrt(inputSize_));

        W1_.resize(hiddenSize_, std::vector<double>(inputSize_));
        b1_.resize(hiddenSize_, 0.0);
        W2_.resize(outputSize_, std::vector<double>(hiddenSize_));
        b2_.resize(outputSize_, 0.0);

        for (auto& row : W1_)
            for (auto& val : row)
                val = dist(gen);

        for (auto& row : W2_)
            for (auto& val : row)
                val = dist(gen);
    }

    void fit(const std::vector<std::vector<double>>& X, const std::vector<int>& y) {
        size_t m = X.size();

        for (int epoch = 0; epoch < epochs_; ++epoch) {
            double lossSum = 0.0;

            for (size_t i = 0; i < m; ++i) {
                // Forward pass
                auto z1 = matVecMul(W1_, X[i], b1_);
                auto a1 = sigmoidVec(z1);
                auto z2 = matVecMul(W2_, a1, b2_);
                auto a2 = sigmoidVec(z2);

                // Compute binary cross-entropy loss
                double y_hat = a2[0];
                int label = y[i];
                double loss = -label * std::log(y_hat + 1e-8) - (1 - label) * std::log(1 - y_hat + 1e-8);
                lossSum += loss;

                // Backpropagation
                double dz2 = y_hat - label; // output error
                std::vector<double> dW2(hiddenSize_);
                for (int j = 0; j < hiddenSize_; ++j)
                    dW2[j] = dz2 * a1[j];

                // Update W2 and b2
                for (int j = 0; j < hiddenSize_; ++j)
                    W2_[0][j] -= learningRate_ * dW2[j];
                b2_[0] -= learningRate_ * dz2;

                // Hidden layer error
                std::vector<double> dz1(hiddenSize_);
                for (int j = 0; j < hiddenSize_; ++j)
                    dz1[j] = sigmoidDeriv(z1[j]) * dz2 * W2_[0][j];

                // Update W1 and b1
                for (int j = 0; j < hiddenSize_; ++j)
                    for (int k = 0; k < inputSize_; ++k)
                        W1_[j][k] -= learningRate_ * dz1[j] * X[i][k];
                for (int j = 0; j < hiddenSize_; ++j)
                    b1_[j] -= learningRate_ * dz1[j];
            }

            if (epoch % 100 == 0) {
                std::cout << "Epoch " << epoch << ", Loss: " << lossSum / m << std::endl;
            }
        }
    }

    std::vector<int> predict(const std::vector<std::vector<double>>& X) const {
        std::vector<int> results;
        for (const auto& x : X) {
            auto a2 = forward(x);
            results.push_back(a2[0] > 0.5 ? 1 : 0);
        }
        return results;
    }

private:
    int inputSize_, hiddenSize_, outputSize_, epochs_;
    double learningRate_;
    std::vector<std::vector<double>> W1_, W2_;
    std::vector<double> b1_, b2_;

    std::vector<double> forward(const std::vector<double>& x) const {
        auto z1 = matVecMul(W1_, x, b1_);
        auto a1 = sigmoidVec(z1);
        auto z2 = matVecMul(W2_, a1, b2_);
        return sigmoidVec(z2);
    }

    static std::vector<double> matVecMul(const std::vector<std::vector<double>>& W,
                                         const std::vector<double>& x,
                                         const std::vector<double>& b) {
        std::vector<double> result(W.size(), 0.0);
        for (size_t i = 0; i < W.size(); ++i) {
            for (size_t j = 0; j < x.size(); ++j)
                result[i] += W[i][j] * x[j];
            result[i] += b[i];
        }
        return result;
    }

    static double sigmoid(double z) {
        return 1.0 / (1.0 + std::exp(-z));
    }

    static double sigmoidDeriv(double z) {
        double s = sigmoid(z);
        return s * (1 - s);
    }

    static std::vector<double> sigmoidVec(const std::vector<double>& z) {
        std::vector<double> result(z.size());
        for (size_t i = 0; i < z.size(); ++i)
            result[i] = sigmoid(z[i]);
        return result;
    }
};

} // namespace models
} // namespace blahajpi
